### PR TITLE
Fix for AllUsesOfAllSymbols missing when using change notifications

### LINKED ISSUE #15960: Getting nothing from GetAllUsesOfAllSymbols() when using DocumentSource.Custom
Hey everyone hello @0101,

we're facing an issue with the checker API:

In tests, the `DocumentSource.Custom` is often used when creating an `FSharpChecker`.
Doing something like:
```fsharp
let checkProjectResults = checker.ParseAndCheckProject(opts) |> Async.RunSynchronously
let allSymbolUses = checkProjectResults.GetAllUsesOfAllSymbols()
```
yields no results in `allSymbolUses`.

It seems to be a caching issue, as the usage of `DocumentSource.Custom` means that `useChangeNotifications` will be set to true in the builder which affects the [slots](https://github.com/dotnet/fsharp/blob/1f37bd6a4df33e01036511856c82ec0cd32408a9/src/Compiler/Service/IncrementalBuild.fs#L967).

A workaround is calling `FSharpChecker.NotifyFileChanged` for the files in the project _before_ calling `ParseAndCheckProject()`.

Is this really the expected behaviour when using `DocumentSource.Custom` or something we should document?

### PR DESCRIPTION
**Pull request by 0101**
Fixes #15960

### CODE REVIEW

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs [hunk 1]
```diff
@@ -132,3 +139,34 @@ let ``Using getSource and notifications instead of filesystem`` () =
...
+            // We should not have to do this:
+            //for p, f in project.GetAllFiles() do
+            //    let fn = p.ProjectDir ++ f.FileName
+            //    do! checker.NotifyFileChanged(fn, options) 
+            let! checkProjectResults = checker.ParseAndCheckProject(options) 
+            return checkProjectResults.GetAllUsesOfAllSymbols()
+        } |> Async.RunSynchronously
+
+
+    traceProvider.ForceFlush() |> ignore
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs:169]: Why this?

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs [hunk 2]
```diff
@@ -132,3 +139,34 @@ let ``Using getSource and notifications instead of filesystem`` () =
...
+
+    use _ = Activity.start "GetAllUsesOfAllSymbols" [  ]
+    
+    let result = 
+        async { 
+            let project = makeTestProject()
+            let checker = ProjectWorkflowBuilder(project, useGetSource=true, useChangeNotifications = true).Checker
+            do! saveProject project false checker 
+            let options = project.GetProjectOptions checker
+            // We should not have to do this:
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs:160]: So can be removed?

**Review by psfinaki — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs (hunk 2 above)
Comment (0101) [reply tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs:160]: Sure.

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs (hunk 1 above)
Comment (0101) [reply tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs:169]: I copy pasted the code from ProjectGeneration. Don't remember why I put it there... either I wasn't always getting the traces or I copy pasted it from somewhere else even then 😅

**Review by T-Gro — APPROVED** [review]

**Pull request merged**